subsys/net/ip/net_if.c

```diff
@@ -1296,44 +1277,185 @@ static void remove_prefix_addresses(struct net_if *iface,
 	}
 }
 
-static inline void prefix_lf_timeout(struct k_work *work)
+static void prefix_lifetime_expired(struct net_if_ipv6_prefix *ifprefix)
 {
-	struct net_if_ipv6_prefix *prefix =
-		CONTAINER_OF(work, struct net_if_ipv6_prefix, lifetime);
 	struct net_if_ipv6 *ipv6;
 
 	NET_DBG("Prefix %s/%d expired",
-		net_sprint_ipv6_addr(&prefix->prefix), prefix->len);
+		net_sprint_ipv6_addr(&ifprefix->prefix), ifprefix->len);
 
-	prefix->is_used = false;
+	ifprefix->is_used = false;
 
-	if (net_if_config_ipv6_get(prefix->iface, &ipv6) < 0) {
+	if (net_if_config_ipv6_get(ifprefix->iface, &ipv6) < 0) {
 		return;
 	}
 
 	/* Remove also all auto addresses if the they have the same prefix.
 	 */
-	remove_prefix_addresses(prefix->iface, ipv6, &prefix->prefix,
-				prefix->len);
+	remove_prefix_addresses(ifprefix->iface, ipv6, &ifprefix->prefix,
+				ifprefix->len);
+
+	net_mgmt_event_notify(NET_EVENT_IPV6_PREFIX_DEL, ifprefix->iface);
+}
+
+static void prefix_timer_remove(struct net_if_ipv6_prefix *ifprefix)
+{
+	NET_DBG("IPv6 prefix %s/%d removed",
+		net_sprint_ipv6_addr(&ifprefix->prefix),
+		ifprefix->len);
+
+	ifprefix->lifetime.timer_timeout = 0;
+	ifprefix->lifetime.wrap_counter = 0;
+
+	sys_slist_find_and_remove(&active_prefix_lifetime_timers,
+				  &ifprefix->lifetime.node);
+}
+
+static bool prefix_manage_timeout(struct net_if_ipv6_prefix *ifprefix,
+				  u32_t current_time, u32_t *next_wakeup)
+{
+	if (check_timeout(ifprefix->lifetime.timer_start,
+			  ifprefix->lifetime.timer_timeout,
+			  ifprefix->lifetime.wrap_counter,
+			  current_time)) {
+		prefix_lifetime_expired(ifprefix);
+		return true;
+	}
+
+	if (current_time == NET_TIMEOUT_MAX_VALUE) {
+		ifprefix->lifetime.wrap_counter--;
+	}
+
+	if (ifprefix->lifetime.wrap_counter > 0) {
+		*next_wakeup = NET_TIMEOUT_MAX_VALUE;
+	} else {
+		*next_wakeup = ifprefix->lifetime.timer_timeout;
+	}
+
+	return false;
+}
+
+static void prefix_lifetime_timeout(struct k_work *work)
```

Comment:
If we have standard structure of struct net_timeout, why do we have duplicate funcs like this for prefixes, addresses?

Reply:
That can be done later if needed. I decided not to do it in this step as that looked quite complicated and more difficult to review.